### PR TITLE
#159: Belos: Provide `Tpetra` version of GCRODR examples

### LINKED ISSUE #159: Belos: Provide `Tpetra` version of GCRODR examples
Child issue of #158

### PR DESCRIPTION
**Pull request by cwschilly**
Fixes #159

### CODE REVIEW

**Comment by cwschilly** [comment]
@stmcgovern This PR is ready for review; I did not convert the preconditioned version, since it would have an Ifpack2 dependency

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    }
+  }
+  TEUCHOS_STANDARD_CATCH_STATEMENTS(verbose, std::cerr, success);
+
+  return success ? EXIT_SUCCESS : EXIT_FAILURE;
+  }
+
+int main(int argc, char *argv[]) {
+  // run with different ST
+  run<double>(argc,argv);
```
Comment (antoinemeyer5) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: need `return` statement

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+template <typename ScalarType>
+int run(int argc, char *argv[]) {
+
+  using ST = typename Tpetra::Vector<ScalarType>::scalar_type;
+  using LO = typename Tpetra::Vector<>::local_ordinal_type;
+  using GO = typename Tpetra::Vector<>::global_ordinal_type;
+  using NT = typename Tpetra::Vector<>::node_type;
+
+  using SCT = typename Teuchos::ScalarTraits<ST>;
+  using MT  = typename SCT::magnitudeType;
```
Comment (antoinemeyer5) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: Maybe we can do directly `using MT = typename Teuchos::ScalarTraits<ST>::magnitudeType;` as `SCT` is used nowhere else.

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+      std::cout << Belos::Belos_Version() << std::endl << std::endl;
+    }
+    //
+    // Get the problem
+    //
+    Belos::Tpetra::HarwellBoeingReader<tcrsmatrix_t> reader( Comm );
+    RCP<tcrsmatrix_t> A = reader.readFromFile( filename );
+    RCP<const tmap_t> Map = A->getDomainMap();
+
+    // Create initial vectors
```
Comment (antoinemeyer5) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp:144]: I would try to be consistent with comments.
Sometimes it's

`// [Your comments]`

and sometimes it's 

```
//
// [Your comments]
//
```

I don't know what is the better way to do it or what the convention is in the Belos package. Personally I think that 1 line `//` is enough.

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp (hunk 1 above)
Comment (cwschilly) [reply packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: Resolved

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp (hunk 3 above)
Comment (cwschilly) [reply packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp:144]: Fixed; I went with the single `//`

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp (hunk 2 above)
Comment (cwschilly) [reply packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: Makes sense to me. Resolved.

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+  TEUCHOS_STANDARD_CATCH_STATEMENTS(verbose, std::cerr, success);
+
+  return success ? EXIT_SUCCESS : EXIT_FAILURE;
+  }
+
+int main(int argc, char *argv[]) {
+  // run with different ST
+  run<double>(argc,argv);
+  // run<float>(argc,argv); // FAILS
+  return 0;
```
Comment (antoinemeyer5) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: I was thinking about something more like:

```
return run<double>(argc, argv);
```

To get the result of the execution/method `run`

Comment (cwschilly) [reply packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: Resolved

**Review by stmcgovern — DISMISSED** [review]
Since eventually testing with multiple scalar types will be adopted. I would prefer 
```
// run with different ST
  run<double>(argc,argv);
  // run<float>(argc,argv); // FAILS
  run<whatever>(argc, argc);
  return 0;
 ```
 as the way to return from main. Since only double is being used now. This style is acceptable to me.

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+  using MVT = typename Belos::MultiVecTraits<ST,MV>;
+  using OPT = typename Belos::OperatorTraits<ST,MV,OP>;
+
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  using Teuchos::ParameterList;
+
+  Teuchos::GlobalMPISession mpiSession (&argc, &argv, &std::cout);
+  const auto Comm = Tpetra::getDefaultComm();
+  const int MyPID = Comm->getRank();
```
Comment (tlamonthezie) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp:97]: in my opinion it is better if variables are camelCase (first letter lowercase)

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+
+  bool verbose = false;
+  bool success = true;
+
+  try {
+    bool proc_verbose = false;
```
Comment (tlamonthezie) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp:103]: Same comment

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+#include <Tpetra_Map.hpp>
+#include <Tpetra_Core.hpp>
+#include <Tpetra_Vector.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+
+// Belos
+#include "BelosConfigDefs.hpp"
+#include "BelosGCRODRSolMgr.hpp"
+#include "BelosLinearProblem.hpp"
+#include "BelosTpetraTestFramework.hpp"
```
Comment (tlamonthezie) [inline packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: Belos include might be after the external includes

Comment (cwschilly) [reply packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp]: Resolved

File: packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp (hunk 5 above)
Comment (cwschilly) [reply packages/belos/tpetra/example/GCRODR/GCRODRTpetraExFile.cpp:97]: These variables were carried over from the epetra versions--they seem to be used in a lot of tests/examples so I think it is easiest to keep them as they are

**Review by stmcgovern — APPROVED** [review]
LGTM

**Pull request merged**